Prototype async REPL using IPython, take III

Carreau opened the pull request:
This is a squash and a rebase of a large number of commits from Min and
I. For simplicity of managing it, history has been reduced to a single
commit, but more historical versions can be found, in particular in
PR #11155, or commit aedb5d6d3a441dcdb7180ac9b5cc03f91329117b to be more
exact.

-- 

I'm expecting some errors in the squash and resolve of conflicts, so will iterate on it.

Carreau committed:
Prototype async REPL using IPython, take III

This is a squash and a rebase of a large number of commits from Min and
I. For simplicity of managing it, history has been reduced to a single
commit, but more historical versions can be found, in particular in
PR 11155, or commit aedb5d6d3a441dcdb7180ac9b5cc03f91329117b to be more
exact.

Carreau committed:
fix UnboundLocalError

Carreau committed:
titleto have doc build

Carreau committed:
fix docs

Carreau committed:
Load the asycn ext only on 3.5+

Carreau committed:
fix runnign on 3.7

Carreau commented:
I'm thinking of dropping 3.4, as maintaining compat with it will be painful:

> There are no specific plans for the next release of Python 3.4. However, the expectation is that Python 3.4.10 will be released in March of 2019, and this will be the final release of Python 3.4.

We'll also be supporting 3.5, 3.6 and 3.7.

@minrk  and @takluyver thoughts ?

minrk commented:
asyncio support is pretty valuable and long overdue, so if 3.4 makes it too hard, I think the trade-off is worth it.

minrk commented on setup.py:
```diff
@@ -201,6 +201,7 @@
 
 extras_require.update({
     ':python_version == "3.4"': ['typing'],
+    ':python_version >= "3.5"': ['trio', 'curio'],
```
Neither of these are dependencies, right? We shouldn't require these packages.

Carreau replied:
One of them is necessary if we want to have  nested `embed()`.  I can make that optional, I probably added them here to have the test passing.

Carreau committed:
remove duplicate WatsNew from bad rebase

Carreau committed:
test with trio

Carreau committed:
3.7 still nto on travis

Carreau committed:
reformat with black

Carreau committed:
doc, remove  appveyor 3.4

Carreau committed:
move infor to the right place

Carreau committed:
DeprecationWarning

Carreau committed:
runblack on new tests

Carreau committed:
att triio runner at top level now that 3.4 drop + docs

Carreau committed:
Add pseudo sync mode

Carreau committed:
no f-strings

dalejung committed:
Only trigger async loop runner when needed. Allows running files that
call async loops synchronously on their own.
https://github.com/ipython/ipython/pull/11265

Handle early return conditions from coro

dalejung commented:
@Carreau I ran into an issue running python files that have their own loop calls. So if I have a file that calls `run_until_complete`, it'll break when %run since it nests within the loop_runners own call.

Can we change the run_cell_async coro to first yield interactivity so it can call the pseudo sync when async is not required?

https://github.com/dalejung/ipython/commit/8d38b3ce41953d8d9048092cc283eb7662a768ed

is a working example. `await` statements in the shell work while running or calling funcs/files that invoke their own loops work as well.

Carreau commented:
I have to better understand the implication. It is likely possible, though it will likely work only on terminal IPython. 

It might also not stay forever, as in the end I want IPython to always run on an event loop. Though, I guess we could run only on the pseudo-sync, or a non-asyncio one. 

Once that hapend we can figure out how to patch `%run` in order to patch `run_until_complete`

njsmith commented:
The way I would imagine it working in the long run is:

* The `async` annotations spread out from `run_cell` to infect the rest of IPython, so that the whole program runs under a single call to the event loop
* By default IPython starts out in 'sync' mode, which means:
  * Using the pseudo-sync runner
  * Raise `SyntaxError` if someone tries to use a top-level `await` in a cell
* You can switch from 'sync' mode to 'asyncio' mode, 'trio' mode etc., which switches IPython to using these loops
  * Internally, this loop switcheroo can be accomplished by yielding a special value to the pseudo-sync runner, which then resumes the same coroutine object under the new loop. This is serious black magic but should work...
  * Also it sets a flag internally to start allowing top-level `await` inside cells
* You can use configuration, pass a command-line flag, etc., to preemptively switch into one of the async modes at startup

Of course there's a ways to go before we use async annotations everywhere like that. But we need to eventually reach a point of keeping the same loop alive over multiple cells, if we want this to be really useful. In any case, I think the basic idea of tracking whether we're in sync or async mode, and turning the magic parser on-or-off as appropriate, is a good one to expose to the user. And that's something that can be done now and would solve @dalejung's problem.

njsmith commented:
...of course when writing the above I forgot about prompt-toolkit. Besides it needing some work to integrate with all these different loops, I'm not sure that it'll be happy switching between loops mid-stream...

Well, the architecture I described above ought to work for the kernel, or for `jupyter console`, at least :-). (With no GUI integration enabled.)

Carreau commented:
> I forgot about prompt-toolkit

Yep, that was one of my main concern, but we can see once we're there. Seeing that it took us 2 years between prototype and now, maybe by then code Python will have a better solution for `async-exec` and multiple loop.

tacaswell commented:
> (With no GUI integration enabled.)

GUI integration is pretty key....

The way the input hooks are implemented now they are very much preemptive rather than cooperative which seems like a hard square to fit into a round hole. 

The way we deal with this at my day-job is having a self-re-triggering task that exhausts the GUI task queue and then schedules its self for some small time in the future. 

@Carreau How through of a review do you want on this (on a scale of "theory sounds good" to "picky implementation")?

pganssle commented on IPython/core/interactiveshell.py:
```diff
@@ -258,6 +358,43 @@ class InteractiveShell(SingletonConfigurable):
+        """
+    ).tag(config=True)
+
+    loop_runner_map ={
```
Note: For another PR, it would be a good idea to have a way to register new names for this.

pganssle commented on docs/source/interactive/autoawait.rst:
```diff
@@ -0,0 +1,210 @@
+    In [1]: %autoawait False
+
+    In [2]: %autoawait
+    IPython autoawait is `Off`, and set to use `IPython.core.interactiveshell._asyncio_runner`
```
This seems like it is exposing an implementation detail. it would probably be worth defining a `__str__` or something on the runners so that they reflect more closely how they were activated.

pganssle commented:
Testing this out, it seems like this breaks `%paste`, even without autoawait on:

```python
$ ipython
Python 3.7.0 (default, Jul 30 2018, 14:36:22) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.0.0.dev -- An enhanced Interactive Python. Type '?' for help.

In [1]: %paste                                                                                                                              
def parseprint(code, filename="<string>", mode="exec", **kwargs):
    """Parse some code from a string and pretty-print it."""
    node = parse(code, mode=mode)   # An ode to the code
    print(dump(node, **kwargs))
    
def pn(tree):
    print(dump(tree))

## -- End pasted text --
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-cda6375ef3e5> in <module>
----> 1 get_ipython().run_line_magic('paste', '')

....

    509         if self.is_running():
--> 510             raise RuntimeError('This event loop is already running')
    511         if events._get_running_loop() is not None:
    512             raise RuntimeError(

RuntimeError: This event loop is already running
```

Carreau commented:
@dalejung  I've added your commit to this pull request, it does fix @pganssle use case as well as %paste and %cpaste were internally running `run_cell`, which was thus nested.

pganssle committed:
Improve async detection mechanism with blacklist

Because the async repl works by wrapping any code that raises
SyntaxError in an async function and trying to execute it again, cell
bodies that are invalid at the top level but valid in functions and
methods (e.g. return and yield statements) currently allow executing
invalid code.

This patch blacklists return and yield statements outside of a function
or method to restore the proper SyntaxError behavior.

pganssle committed:
Add tests for SyntaxError with top-level return

pganssle commented on IPython/core/tests/test_async_helpers.py:
```diff
@@ -0,0 +1,237 @@
+            iprc(
+                """
+                from asyncio import sleep
+                await.sleep(0.1)
```
This is an example of where the test isn't working right - `await.sleep` is incorrect.

That said, it turns out that:

`iprc("""
from asyncio import sleep
await sleep(0.1)
""")

is *also* a `SyntaxError`, for inscrutable reasons.

pganssle replied:
Found the problem:

```python
In [1]: from IPython.core.async_helpers import _asyncify                                                                                    
In [2]: from textwrap import dedent
In [3]: print(_asyncify(dedent(""" 
   ...: import asyncio 
   ...: await sleep(0.001) 
   ...: """)))                                                                                                                              

async def __wrapper__():
    try:
         import asyncio
        await sleep(0.001)

    finally:
        locals()
```

Note the erroneous indentation on the `from asyncio`. I think `_asyncify` is transforming the `\n` into a space somewhere.

pganssle committed:
Fix indentation problem with _asyncify

If the input to `_asyncify` contains newlines, the strategy of indenting
and then truncating the first 8 characters won't work, since the first,
empty line doesn't get indented. Instead I've changed the format string
to accept the normal output of textwrap.indent() without modification.

pganssle committed:
Fix async_helpers tests

Many of the tests were simply executing code and ignoring errors
(including several which had errors in them). This commit makes it so
that errors are actually raised, and fixes the errors in code that had
them.

pganssle commented:
@Carreau I've pushed two commits to the branch adding support for properly detecting syntax errors.

Also, it seems that the tests weren't actually failing if the test failed, because errors weren't being raised, and indeed several of the tests *should* have been failing. I've pushed two more commits fixing the tests and a bug in `_asyncify` that was causing test failures.

pganssle committed:
Refactor and tweak async-in-function tests

This improves Python3.5 compatibility, since "yield" in an async
function was a SyntaxError in 3.5, but not in later versions

Also adds a test for async methods.

Carreau committed:
docs cleanup, reformat code, remove dead code.

Carreau committed:
remove trio from test requirement and skipp if not there

Carreau committed:
fix test

Carreau commented:
Todo, catch exception when the runner crash, EG, when using `autoawait trio`, and then `await asyncio.sleep(0)`.

Carreau committed:
Test that using wrong runner/coroutine pair does not crash.

Make sure that if the coroutine runner encounter an exception (likely
when encountering the wrong coroutine from another aio library), it is
set on the result and does not crash the interpreter.

Carreau committed:
run tests with trio and curio on travis

Carreau commented:
Timeit, time, and prun magics will need updates to work with async code. 
They still work for sync code though, so I don't think it is a blocker.

Carreau committed:
add notes some magic don't work yet

Carreau committed:
typo

Carreau commented on docs/source/whatsnew/pr/deprecations.rst:
```diff
@@ -1,3 +1,6 @@
+Depreations
```
Thanks, Fixed.

minrk commented on IPython/core/interactiveshell.py:
```diff
@@ -2780,9 +2984,14 @@ def error_before_exec(value):
+                    interactivity = 'async'
+                # yield interactivity so let run_cell decide whether to use
+                # an async loop_runner
+                yield interactivity
```
Adding `yield interactivity` here makes this a not-entirely-valid coroutine, doesn't it? When I try to call `await run_cell_async` from ipykernel, I get a `bad yield: 'last_expr'`. How can we communicate this required info without faking that this isn't actually a coroutine? Does it need to be an actual async generator?

Carreau commented:
That's right. You need to do a small dance. I believe I have a patch on my
machine for ipykernel to make this work.
You need to advance the coro by sending None to it once. We could do it in
another way, but at some point we'll always hit the fact that python does
not have a real async-exec and that we need to distinguish 2 path of
running sync code, or asyc one.

On Tue, Sep 4, 2018, 16:43 Min RK <notifications@github.com> wrote:

> *@minrk* commented on this pull request.
> ------------------------------
>
> In IPython/core/interactiveshell.py
> <https://github.com/ipython/ipython/pull/11265#discussion_r214942959>:
>
> > @@ -2780,9 +2984,14 @@ def error_before_exec(value):
>                  self.displayhook.exec_result = result
>
>                  # Execute the user code
> -                interactivity = 'none' if silent else self.ast_node_interactivity
> -                has_raised = self.run_ast_nodes(code_ast.body, cell_name,
> -                   interactivity=interactivity, compiler=compiler, result=result)
> +                interactivity = "none" if silent else self.ast_node_interactivity
> +                if _run_async:
> +                    interactivity = 'async'
> +                # yield interactivity so let run_cell decide whether to use
> +                # an async loop_runner
> +                yield interactivity
>
> Adding yield interactivity here makes this a not-entirely-valid
> coroutine, doesn't it? When I try to call await run_cell_async from
> ipykernel, I get a bad yield: 'last_expr'. How can we communicate this
> required info without faking that this isn't actually a coroutine? Does it
> need to be an actual async generator?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ipython/ipython/pull/11265#pullrequestreview-152101058>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUez6-l5irZuy9QnGMdamhrYJsfea77ks5uXpGXgaJpZM4V7qox>
> .
>

Carreau committed:
Tool to go around sphinx limitation during developpement

Carreau committed:
run doc fixing tool on travis

Carreau commented:
Ok, seem like we've reached quasi static state and most of the latest updates were documentation. Plus ipykernel works with that. So I'm going to merge (thanks all involved !), and we can refine if necessary.

minrk commented:
#11289 makes run_cell_async a regular coroutine by moving the 'should it be async' check to a separate method.

Since in general, running it with a coroutine runner ought to work (I know there are edge cases for now), skipping this check should be safe.
